src/aosm/azext_aosm/custom.py

```diff
@@ -0,0 +1,527 @@
+# --------------------------------------------------------------------------------------------
+
+import json
+import os
+import shutil
+from dataclasses import asdict
+from pathlib import Path
+from typing import Optional
+
+from azure.cli.core.azclierror import (
+    CLIInternalError,
+    InvalidArgumentValueError,
+    UnclassifiedUserFault,
+)
+from azure.cli.core.commands import AzCliCommand
+from azure.core import exceptions as azure_exceptions
+from knack.log import get_logger
+
+from azext_aosm._client_factory import cf_features, cf_resources
+from azext_aosm._configuration import (
+    CNFConfiguration,
+    Configuration,
+    NFConfiguration,
+    NSConfiguration,
+    VNFConfiguration,
+    get_configuration,
+)
+from azext_aosm.delete.delete import ResourceDeleter
+from azext_aosm.deploy.deploy_with_arm import DeployerViaArm
+from azext_aosm.generate_nfd.cnf_nfd_generator import CnfNfdGenerator
+from azext_aosm.generate_nfd.nfd_generator_base import NFDGenerator
+from azext_aosm.generate_nfd.vnf_nfd_generator import VnfNfdGenerator
+from azext_aosm.generate_nsd.nsd_generator import NSDGenerator
+from azext_aosm.util.constants import (
+    AOSM_FEATURE_NAMESPACE,
+    AOSM_REQUIRED_FEATURES,
+    CNF,
+    NSD,
+    VNF,
+    DeployableResourceTypes,
+    SkipSteps,
+)
+from azext_aosm.util.management_clients import ApiClients
+from azext_aosm.vendored_sdks import HybridNetworkManagementClient
+
+logger = get_logger(__name__)
+
+
+def build_definition(
+    definition_type: str,
+    config_file: str,
+    order_params: bool = False,
+    interactive: bool = False,
+    force: bool = False,
+):
+    """
+    Build a definition.
+
+    :param definition_type: VNF or CNF
+    :param config_file: path to the file
+    :param definition_type: VNF, CNF
+    :param interactive - whether to prompt for input when creating deploy parameters
+                         mapping files
+    :param force: force the build even if the design has already been built
+    """
+
+    # Read the config from the given file
+    config = _get_config_from_file(
+        config_file=config_file, configuration_type=definition_type
+    )
+    assert isinstance(config, NFConfiguration)
+
+    # Generate the NFD and the artifact manifest.
+    _generate_nfd(
+        definition_type=definition_type,
+        config=config,
+        order_params=order_params,
+        interactive=interactive,
+        force=force,
+    )
+
+
+def generate_definition_config(definition_type: str, output_file: str = "input.json"):
+    """
+    Generate an example config file for building a definition.
+
+    :param definition_type: CNF, VNF
+    :param output_file: path to output config file, defaults to "input.json"
+    """
+    config: Configuration
+    if definition_type == CNF:
+        config = CNFConfiguration.helptext()
+    elif definition_type == VNF:
+        config = VNFConfiguration.helptext()
+
+    _generate_config(configuration=config, output_file=output_file)
```

Comment:
the `config` variable might be referenced before assignment, since `definition_type` parameter might be None.

Reply:
Fixed